linux_kernel_modules/rtc_sync/rtc_sync.c

```diff
@@ -0,0 +1,176 @@
+
+#include <linux/module.h>
+#include <linux/rtc.h>
+#include <linux/timer.h>
+#include <linux/time.h>
+#include <linux/sched.h>
+#include <linux/kthread.h>
+
+
+static struct task_struct *time_check_thread;
+
+
+//this function syncs system time with RTC when startup - if there is valid time store in RTC
+static int rtc_hctosys(void)
+{
+	int err;
+	struct rtc_time tm;
+	struct rtc_device *rtc = rtc_class_open("rtc1");
+	if (rtc == NULL) {
+		printk("%s: unable to open rtc device (rtc1) for rtc_hctosys\n",
+			__FILE__);
+		return -ENODEV;
+	}
+
+	err = rtc_read_time(rtc, &tm);
+	if (err == 0) {
+		err = rtc_valid_tm(&tm);
+		if (err == 0) {
+			struct timespec tv;
+
+			tv.tv_nsec = NSEC_PER_SEC >> 1;
+
+			rtc_tm_to_time(&tm, &tv.tv_sec);
+
+			do_settimeofday(&tv);
+
+			dev_info(rtc->dev.parent,
+				"setting system clock to "
+				"%d-%02d-%02d %02d:%02d:%02d UTC (%u)\n",
+				tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+				tm.tm_hour, tm.tm_min, tm.tm_sec,
+				(unsigned int) tv.tv_sec);
+		} else { 
+			dev_err(rtc->dev.parent,
+				"hctosys: invalid date/time\n");
+		}
+	} else {
+		dev_err(rtc->dev.parent,
+			"hctosys: unable to read the hardware clock\n");
+	}
+
+	rtc_class_close(rtc);
+
+	return 0;
+}
+
+//systohc sets RTC with system time
+static int systohc(struct timespec now)
+{
+	
+        struct rtc_device *rtc;
+	struct rtc_time tm;
+	int err = -ENODEV;
+	printk("Setting time to %lld\n", timespec_to_ns(&now));
+
+	if (now.tv_nsec < (NSEC_PER_SEC >> 1))
+		rtc_time_to_tm(now.tv_sec, &tm);
+	else
+		rtc_time_to_tm(now.tv_sec + 1, &tm);
+
+	rtc = rtc_class_open("rtc1");
+
+        if (rtc == NULL) {
+		printk("%s: unable to open rtc device (rtc1) for systohc\n",
+			__FILE__);
+		return -ENODEV;
+	}
+
+	if (rtc) {
+		/* rtc_hctosys exclusively uses UTC, so we call set_time here,
+		 * not set_mmss. */
+		if (rtc->ops &&
+		    (rtc->ops->set_time ||
+		     rtc->ops->set_mmss))
+			err = rtc_set_time(rtc, &tm);
+		rtc_class_close(rtc);
+	}
+        
+	return 0;
+}
+
+//time_check chekcs if the system time has been modified during the last checking period by comparing the expected time after checking period with a time allowance
+static int time_check(void *data)
+{
+        //checking period
+        struct timespec delta_expected = { .tv_sec=10 };
+        //time allowance
+	struct timespec delta_delta_acceptable = { .tv_sec=1 };
+        long long allowance = timespec_to_ns(&delta_delta_acceptable);
+
+	while (!kthread_should_stop()) {
+		struct timespec before;
+		struct timespec after;
+        	struct timespec delta_actual;
+        	long long timelapse;
+		long long timediff;
+		long timeout = 10 * HZ;
+
+		ktime_get_real_ts(&before);
+
+		while (timeout > 0)
+			timeout = schedule_timeout(timeout);
+		ktime_get_real_ts(&after);
+
+		//compare current time with previous
+		delta_actual = timespec_sub(after, before);
+		timelapse = timespec_to_ns(&delta_actual);
+		//absolute value of timelapse
+		if (timelapse <= 0)
```

Comment:
You can use `abs64(val)`  to get the absolute value.  So for example you chould change this to `timelapse = abs64(timespec_to_ns(&delta_actual));`